straxen/corrections_services.py

```diff
@@ -327,6 +326,25 @@ def get_cmt_local_versions(global_version):
     return cmt.get_local_versions(global_version)
 
 
+args_idx = lambda x: x.rfind('?') if '?' in x else None
+
+
+def replace_version(url, version):
+    kwargs = {k:v[0] for k,v in parse_qs(urlparse(url).query).items()}
+    kwargs['version'] = version
+    args = [f"{k}={v}" for k,v in kwargs.items()]
+    args_str = "&".join(args)
+    return f'{url[:args_idx(url)]}?{args_str}'
+
+
+def correction_name_from_urlstring(url):
+    # only care about what happens after cmt
+    before_url, cmt, after_url = url.partition('cmt://')
+    tmp_config = straxen.URLConfig(default=after_url)
+    new_url, kwargs = tmp_config.split_url_kwargs(after_url)
+    return tmp_config.dispatch(new_url, **kwargs)
```

Comment:
nice hack! any reason you decided to put this here and not in the `get_cmt_options()` function? in that function you already have access to the URLConfig instance so you can skip creating a new one. Then you already return the actual names for anyone else that may be using `get_cmt_options()` in their code, no only this script for global_versions...
The only limitation in this hack is that now you cannot have the cmt option you use depend on the run_id or any other run-time configured things. I guess that functionality is not really needed in 99.99% of cases though, plus this is only temporary as CMT is upgraded.

Reply:
Minor comment can you put this below `apply_cmt_version`. In the beginning it was not entirely clear to me why I am reading it. I think in general code should always flow from general to detail.

Reply:
@jmosbacher, thanks that's a good suggestion. I will move

Reply:
@jmosbacher sorry now that I'm looking at it I think I actually am not sure what you suggest. 

`get_cmt_options` is getting the strax option/config name and the values for all options that use CMT. So this itself does not care what the actual CMT collection name is (though ideally these things would be the same, they are not).

So what do we gain by adding this to `get_cmt_options` instead of here in `apply_cmt_version` ?

Reply:
I guess maybe we can change what this `get_cmt_options` returns? It could be a nested dict that returns something like 

```
{'strax_option`: {'collection_name': 'some_name',
                             'version': 'some_local_version'
                            }
}
```

Reply:
Sorry for not being more explicit. 
What i had in mind was to have `get_cmt_options` return a tuple consistent with the cmt structure so first item is the name and second is maybe the URL string. but i guess its not very crucial, i doubt anyone else has written any code using `get_cmt_options`

Reply:
but not sure its worth the effort

Reply:
The way it works now is that it returns a dict like this: 

{'strax_option1': (some_cmt_tuple, if_not, using_urlconfig),
 'strax_option2': 'protocol://could_also_be_a_urlconfig'
}

So what the function in question is doing is grabbing the collection name (which is in CMT tuples already) from the URL string. 

So I still don't understand what the suggestion is, sorry 😕

Reply:
I was thinking something like this:

```python
def get_cmt_options(context):
    """
    Function which loops over all plugin configs and returns dictionary
    with option name as key and current settings as values.

    :param context: Context with registered plugins.
    """
    cmt_options = {}
    for data_type, plugin in context._plugin_class_registry.items():
        for option_key, option in plugin.takes_config.items():
            value = option.default
            if option_key in context.config:
                value = context.config[option_key]

            if isinstance(option, URLConfig) and 'cmt://' in value:
                _, _, after_url = value.partition('cmt://')
                new_url, kwargs = option.split_url_kwargs(after_url)
                cmt_name =  option.dispatch(new_url, **kwargs)
                value = (cmt_name, value, True)

            if straxen.get_corrections.is_cmt_option(value):
                cmt_options[option_key] = value
            
           
    return cmt_options

```

then we dont have to touch the `_is_cmt_option` function, so less chance of breaking any existing code somewhere.
 and int the just have to check if the version part of the `apply_cmt_version` function we just check if `cmt://` is in `tup[1]` to know if its a urlconfig and treat it accordingly.

Reply:
Thanks for the clear example @jmosbacher. Personally I would prefer to keep the `_is_cmt_option` how we have it now because that can be easily applied to both CMT tuple configs and url configs. What you propose would return things like this, correct? 

```
{'strax_option1': ('collection_name', 'collection_version', True),  # a tuple config 
  'strax_option2' : ('collection_name', 'cmt://collection_name?etc', True)  # url config
}
```

Where the strax_option2 is a bit nonsensical right? Sorry for being so slow/pedantic here, just still not understanding what we gain.

Reply:
Fair enough, to me this just seemed like the natural place to do the lookup for the cmt name. I always hate changing the behavior of an existing function because you never know who uses that function and is relying on the existing behavior in the code.

But I think your implementation is fine, and most importantly it works so I'm fine with it being there until all the configs are converted.